### DISCUSSION TITLE
Z state update clarification in the code (when using flowdeck)

### DISCUSSION THREAD

**Discussion by siddarth-murugesan** [bitcraze/discussions#517] [Q&A]
Hi,

I am using Crazyflie 2.1 with only flow deck v2 (no other decks). I am working on a project trying to maintain a constant altitude Z state (say 1m) throughout the flight between two setpoints even with an obstacle of height 0.25m in the path. I want to avoid drone's height shooting up when crossing over the obstacle thinking the floor has changed and trying to maintain the same height. I have commented out the code in [mm_tof.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/kalman_core/mm_tof.c) and [mm_flow.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/kalman_core/mm_flow.c) where the Z state update cycle happens for flow deck. But even with commenting out the Z state update code from the flow deck, the drone gets to shoot up when flying over the obstacle, thinking the floor has changed and thrusting up to maintain the constant height from the obstacle.

I had almost searched all other places where the Z state updates in the code, but I am unable to find it. Could you please help me, to understand where other places the Z state update could come from the code, or is my approach the right way ahead?

Sorry for the long question and Thanks in advance

**Comment by hmllr** [comment]
It is always good to explain the question in detail, so don't worry about a long question :)

I just tried it out myself to be sure - but commenting out the update in mm_tof.c is enough to not consider the tof measurement anymore.
It should not react to changes on the floor anymore, otherwise it seems as if something with flashing it went wrong (you can see if your firmware is modified in the console of the cfclient). 
You might notice the drone drifting upwards (maybe also downwards, but from experience it is usually upwards) in z direction though - this is because with only the  IMU the position error accumulates (and the barometer is usually not used as it is also not very accurate).
I would recommend to implement a filter to recognize the step (maybe even fuse with the barometer measurements, but I don't think it is really necessary). There exist different complex solutions, from just ignoring some ToF measurements if the error is over a threshold (only for small obstacles) to using a multiranger and the upward facing sensor (https://github.com/bitcraze/crazyflie-firmware/pull/823) to a more elaborate filter taking ToF and IMU into account (as e.g. here: https://github.com/Xeratec/crazyflie-stepStabilizer/blob/main/src/app/src/app_stepStabilizer_filter.c this was used for a ML project, but ML is an overkill for the task and was just used for the exercises sake).
Hope this helps,
Hanna

**Comment by siddarth-murugesan** [comment]
Hi,
Thank you for your quick response. 

1- I did the same as commenting out Z update in mm_tof.c and even in mm_flow.c and it still reacts to obstacles somehow. Could you please tell me how you have checked the code change? For testing the code, I am using motion_commander in python script, setting a default height and take off, move forward, stop commands with an obstacle in the path in the ground.

2- I have uploaded the image of the flashing process below. I think the flashing was fine, as I was able to add some "DEBUG_PRINT" in the code and flash and run it and check its output in the console (commands used are - make cf2_defconfig, make -j 12, make cload)

![image](https://user-images.githubusercontent.com/108895043/212884157-4d0fb587-183c-41e3-8270-35228dc03752.png)

![image](https://user-images.githubusercontent.com/108895043/212884807-d2811a5c-a857-41ae-9c26-6b084fbbf835.png)

3- I could understand that it will drift a bit when commenting out the Z update from mm_tof.c, due to error accumulation of IMU, but here it is still kind of shooting up like 0.2m further high when exactly flying over the obstacle of height 0.2m and coming down. Commenting out the Z update in mm_tof is similar to flying with only the optical movement sensor and without zranger in the flow deck right?

4- Yeah, I tried this code as well (https://github.com/bitcraze/crazyflie-firmware/pull/823), this is where I started and was unable to reproduce the expected result with the given code. When I ran the code, I got the same result as the drone shooting up over the obstacle and as it was an external project, I couldn't get support on it as well 

5- I haven't tried the code from the mentioned ML project yet, will update once I try it

Thanks a lot :)

**Reply by hmllr** [reply]
1) I just took off with the cfclient GUI and then held my hand (very slowly to not make it mess up with the flow measurements too much) under the CF to check that it doesn't jump up. This should do the same thing as you did though.
2) the flashing looks good and the firmware that's running is indeed modified - I don't see your added DEBUG_PRINT, but if you get that it should really have worked. (To be super sure you can run "make clean" before you compile the code, but this should really not be the problem)
3) yes, indeed. It should definitively not shoot up. Yes, it is like flying with only flow measurements and no ToF measurements - the flow measurements are heavily dependent on a accurate height estimation though.
4)  It sounds really weird that it is still shooting up - maybe you can replace the update with a DEBUG_PRINT which says "not doing update" or so? Maybe better only every 10th or so time, as the zranger update runs with 40Hz. This way you would be sure that you changed the code in the right file and are really not doing the update.
5) if you have a flat ceiling I think the approach you tried is better - the ML project I sent you is not maintained and was working with an older firmware, I sent it more for inspiration than for an out of the box solution, I just know it because I supervised it ;)

If nothing helps it might be a good idea to log some variables - like the ToF measurement, state estimate and compare between flying without flow deck, with the zranger update commented in/out, etc., just to understand better what is going on.
